servers/src/grin/sync.rs

```diff
@@ -120,21 +120,51 @@ pub fn run_sync(
 
 					// run the header sync every 10s
 					if si.header_sync_due(&header_head) {
+						header_sync(peers.clone(), chain.clone());
+
 						let status = sync_state.status();
 						match status{
 							SyncStatus::TxHashsetDownload => (),
 							_ => {
-								header_sync(peers.clone(), chain.clone());
 								sync_state.update(SyncStatus::HeaderSync{current_height: header_head.height, highest_height: si.highest_height});
 							}
 						};
 					}
 
 					if fast_sync_enabled {
-						// run fast sync if applicable, every 5 min
-						if header_head.height == si.highest_height && si.fast_sync_due() {
-							fast_sync(peers.clone(), chain.clone(), &header_head);
-							sync_state.update(SyncStatus::TxHashsetDownload);
+
+						// check sync error
```

Comment:
Nitpicking - a fragment from this line to 151 could be a bit simpler like:
```
						let maybe_sync_error = sync_state.sync_error();
						if let Some(ref sync_error) = *maybe_sync_error.read().unwrap() {
							error!(
								LOGGER,
								"fast_sync: error = {:?}. restart fast sync", sync_error
							);
							si.fast_sync_reset();
							sync_state.clear_sync_error();
						}
```

Reply:
Haha~   You're making same error like me:)  Actually, yours (also my previous trying) can't build, because of an odd short live error, plus a dead-lock risk.

Reply:
try it:)

Reply:
I said that's old mistakes what I made before:) it will cause 2 problems there~

Reply:
```
error[E0597]: `maybe_sync_error` does not live long enough
   --> servers/src/grin/sync.rs:139:39
    |
139 |                             if let Some(ref sync_error) = *maybe_sync_error.read().unwrap() {
    |                                                            ^^^^^^^^^^^^^^^^ borrowed value does not live long enough
...
160 |                         }
    |                         - `maybe_sync_error` dropped here while still borrowed
    |
    = note: values in a scope are dropped in the opposite order they are created
```

Reply:
@hashmap   I can add one line after your code to pass the build:
```
drop(maybe_sync_error);
```

But then it's same as my original code:)
And another problem in your code is the `read lock` and the `write lock`, it's a risk to dead lock, I don't know why compiler don't complain about that, but better to avoid.